### DISCUSSION TITLE
How to run gem5 on multiple cores on the host side?

### DISCUSSION THREAD

**Discussion by duqi-hnu** [gem5/gem5#854] [gem5-dev]
Gem5 supports multi-core emulation, but Gem5 itself runs on a single core, no matter how many cores you have on your host. This greatly limits the efficiency of multi-core parallel simulation.

**Comment by OdnetninI** [comment]
As far as I know, you cannot use more than 1 core in the current upstream gem5.
I've seen some forks with parallelism in mind.

However, single-threaded gem5 has some advantages, but the main of them is "Determinism".
It doesn't matter how many times you ran gem5, using the same checkpoint, you always obtain the same result.

I am sure there are more people interested in parallelizing gem5, however, a single-threaded version is much simpler to debug and develop.

In my experience, if your issue is the performance of the simulator, I just run multiple instances with different benchmarks. In the end, it is the same to run 10 simulations in parallel and running 10 simultaneous that are 10x faster.

**Reply by BobbyRBruce** [reply]
Well said! Thanks for your response @OdnetninI !

Just to jump in and re-iterate: The gem5 source has been under continual development for over 20 years and the assumption has always been it runs on a single thread. So there's just a lot of very heavily embedded code that'd need changed for multi-threading to work. If someone wants to have a go at making it multithreaded, please go ahead, but I suspect it'd be difficult and take time to do right (i.e. in a way we can confidently say won't introduce more problems than it's worth).

I personally don't care very much about multithreaded gem5. Most simulation work requires simulating more than one simulation configuration. Therefore, as @OdnetninI said, you can utilize your system resources by running multiple simulations at once. Multiple processes of single-threaded gem5 rather than a single gem5 process running on multiple-threads. Conceptually this is equivalent utilization of hardware, and probably faster than back to back multithreaded application given threads don't scale as well as processes.

**Comment by BobbyRBruce** [comment]
gem5 is single-threaded and there's nothing in the pipeline to change that. In short, its single-threaded nature is heavily imbedded in how gem5 works. Many have attempted to make it a multi-threaded application but none have succeeded. It's just too big of an undertaking.

We generally don't worry about this as most of the times when carrying out research we'll want to run many gem5 simulations in parallel. So, while not multi-threaded, you can have multiple gem5 processes running in parallel to properly utilize your system's resources.

The gem5 Orchestration Framework is going to make handling multiple gem5 processes easier (see Discussion #754).

**Reply by steve-reinhardt** [reply]
Sorry to be so late to the party, but I just wanted to point out that there was some work on multi-threading the simulation a long time ago.  [This commit](https://github.com/gem5/gem5/commit/de366a16f11b7e27a5b5e064a2a773052568428e) is over 10 years old and I think the code changes themselves are even older. It sets up a framework where multiple concurrent threads each have their own event queue, and collections of SimObjects share each event queue. Those different collections of SimObjects can then be simulated in parallel. The challenge is that any communication between collections of objects ends up scheduling a cross-thread event which then requires a thread barrier to enforce determinism, and the frequency of that barrier is a function of the minimum communication latency between collections. So it works OK if the collections of objects represent different network nodes, for instance, but the communication latency between cores in a multi-core CPU is likely too low and results in synchronization that is likely too frequent to provide much speedup. This parallelization is based on some really ancient work [1](https://dl.acm.org/doi/10.1145/166955.166979) [2](https://pages.cs.wisc.edu/~wwt/wwt2).

There's also the [dist-gem5](https://publish.illinois.edu/icsl-pdgem5) work out of Illinois that uses multiple hosts to run simulations of networked systems but I don't know if anyone's actively using that either.

**Reply by andysan** [reply]
There are two main use cases for the threaded gem5 at the moment. The main one is probably KVM where each KVM CPUs together with their private peripherals get pinned to their own threads and system components typically end up in their own thread. KVM is currently not deterministic (it would probably be possible to make it deterministic), so the multi-threaded setup uses relaxed barriers inspired by Graphite.

The other main use case is when simulating multiple systems in the same process. In this case, communication happens across a network (typically Ethernet) which means that the number of places we need to care about locking is limited. An added benefit is that the communication latency is high compared to the rest of the system, so barriers can be relatively sparse and still maintain determinism.

If memory serves me right, the dist-gem5 work out of Illinois and Arm Research is actually subtly different. In this case, we use multiple processes and MPI to simulate multiple systems. This lets you simulate a cluster on a cluster.

The main issue when extending gem5 to multiple threads is the lack of well-defined interfaces between some components. There are multiple places where SimObjects call directly into other SimObjects (e.g., to deliver interrupts). Similarly, the interconnect models aren't thread safe. @tiagormk was looking at this many years ago and I believe he had a working prototype.

**Reply by tiagormk** [reply]
As @andysan mentioned, I made an attempt at it some years ago. At the time I was using the classic caches and there was some benefit to be had for workloads that mostly hit in the private caches, but as soon as snooping is required performance drops.  Snoops propagate atomically and may update the state of multiple caches in a single simulation tick, so any event that can generate snoops cannot execute concurrently with any other event that may touch the caches, which basically serializes the whole simulation.

If anyone wants to tackle this, I'd recommend looking at the Ruby memory system. All components in Ruby communicate only through  MessageBuffer objects, so in theory things should work by only adding local synchronization to those objects. Other things such as interrupts, IO, functional accesses, etc, will probably require some sort of global locking though.
